master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/AbstractMetaManager.java

```diff
@@ -161,7 +162,8 @@ public void updateWorkerHeartbeatMeta(
     }
     appDiskUsageMetric.update(estimatedAppDiskUsage);
     // If using HDFSONLY mode, workers with empty disks should not be put into excluded worker list.
-    if (!excludedWorkers.contains(worker) && (disks.isEmpty() && !conf.hasHDFSStorage())) {
+    if (!excludedWorkers.contains(worker)
+        && ((disks.isEmpty() && !conf.hasHDFSStorage()) || highWorkload)) {
```

Comment:
this may cause HDFS-only workers to never back?

Reply:
HDFS-only workers' available slots are 0. This will cause workers to get excluded forever.

Reply:
fix it

Reply:
> this may cause HDFS-only workers to never back?

Nice catch!